File: backend/routes/diagnostic.js
const express = require('express');
const router = express.Router();
const { exec } = require('child_process');
const net = require('net');
const dns = require('dns');
const util = require('util');
const ping = require('ping');

const dnsLookup = util.promisify(dns.lookup);
const dnsResolve = util.promisify(dns.resolve);

// Middleware de autenticação
const authenticateToken = (req, res, next) => {
  const authHeader = req.headers['authorization'];
  const token = authHeader && authHeader.split(' ')[1];

  if (!token) {
    return res.status(401).json({ error: 'Token não fornecido' });
  }

  jwt.verify(token, JWT_SECRET, (err, user) => {
    if (err) return res.status(403).json({ error: 'Token inválido' });
    req.user = user;
    next();
  });
};

// ==================== 1. PING AVANÇADO ====================
router.post('/ping', authenticateToken, async (req, res) => {
  const { host, count = 5 } = req.body;

  if (!host) {
    return res.status(400).json({ error: 'Host é obrigatório' });
  }

  const latencies = [];
  let packetLoss = 0;
  let successCount = 0;

  for (let i = 0; i < count; i++) {
    try {
      const startTime = Date.now();
      const result = await ping.promise.probe(host, { timeout: 2 });
      const endTime = Date.now();

      if (result.alive) {
        const latency = endTime - startTime;
        latencies.push(latency);
        successCount++;
      }
    } catch (err) {
      console.error('Erro no ping:', err);
    }
    // Aguarda 1 segundo entre pings
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  packetLoss = ((count - successCount) / count) * 100;

  const avgLatency = latencies.length > 0
    ? Math.round(latencies.reduce((a, b) => a + b, 0) / latencies.length)
    : null;
  const minLatency = latencies.length > 0 ? Math.min(...latencies) : null;
  const maxLatency = latencies.length > 0 ? Math.max(...latencies) : null;

  res.json({
    host,
    status: successCount > 0 ? 'online' : 'offline',
    packet_loss: packetLoss,
    avg_latency: avgLatency,
    min_latency: minLatency,
    max_latency: maxLatency,
    success_count: successCount,
    total_count: count
  });
});

// ==================== 2. TRACEROUTE ====================
router.post('/traceroute', authenticateToken, async (req, res) => {
  const { host } = req.body;

  if (!host) {
    return res.status(400).json({ error: 'Host é obrigatório' });
  }

  // Usando comando nativo (Linux/Mac)
  exec(`traceroute -n -m 30 ${host}`, { timeout: 30000 }, (error, stdout, stderr) => {
    if (error && error.code !== 1) {
      // Tentativa alternativa com tcptraceroute ou fallback
      exec(`tracert -d -h 30 ${host}`, { timeout: 30000 }, (err2, stdout2) => {
        if (err2) {
          return res.status(500).json({ error: 'Erro ao executar traceroute' });
        }
        const hops = parseTraceroute(stdout2);
        res.json({ hops, raw: stdout2 });
      });
    } else {
      const hops = parseTraceroute(stdout);
      res.json({ hops, raw: stdout });
    }
  });
});

function parseTraceroute(output) {
  const lines = output.split('\n');
  const hops = [];

  for (const line of lines) {
    const match = line.match(/^\s*(\d+)\s+(\d+\.\d+\.\d+\.\d+|\*+)\s+([\d.]+|\*+)/);
    if (match) {
      hops.push({
        hop: parseInt(match[1]),
        ip: match[2],
        latency: match[3] !== '*' ? parseFloat(match[3]) : null
      });
    }
  }
  return hops;
}

// ==================== 3. TESTE DE PORTA ====================
router.post('/port-check', authenticateToken, async (req, res) => {
  const { host, port, timeout = 3000 } = req.body;

  if (!host || !port) {
    return res.status(400).json({ error: 'Host e porta são obrigatórios' });
  }

  const results = [];

  const checkPort = (p) => {
    return new Promise((resolve) => {
      const socket = new net.Socket();
      const startTime = Date.now();

      socket.setTimeout(timeout);
      socket.connect(p, host, () => {
        const latency = Date.now() - startTime;
        socket.destroy();
        resolve({ port: p, open: true, latency });
      });

      socket.on('error', () => {
        socket.destroy();
        resolve({ port: p, open: false, latency: null });
      });

      socket.on('timeout', () => {
        socket.destroy();
        resolve({ port: p, open: false, latency: null });
      });
    });
  };

  // Suporta array de portas ou porta única
  const ports = Array.isArray(port) ? port : [port];

  for (const p of ports) {
    const result = await checkPort(parseInt(p));
    results.push(result);
  }

  res.json({ host, results });
});

// ==================== 4. DNS LOOKUP ====================
router.post('/dns-lookup', authenticateToken, async (req, res) => {
  const { domain, recordType = 'A' } = req.body;

  if (!domain) {
    return res.status(400).json({ error: 'Domínio é obrigatório' });
  }

  try {
    // Resolve A records
    const addresses = await dnsResolve(domain, recordType);

    // Tenta resolver reversamente (PTR)
    let reverseLookup = null;
    if (addresses && addresses[0]) {
      try {
        reverseLookup = await dnsLookup(addresses[0]);
      } catch (err) {
        // Ignora erro de reverse lookup
      }
    }

    res.json({
      domain,
      record_type: recordType,
      records: addresses.map(addr => ({ value: addr })),
      reverse_lookup: reverseLookup?.hostname || null,
      success: true
    });
  } catch (error) {
    res.json({
      domain,
      record_type: recordType,
      records: [],
      error: error.message,
      success: false
    });
  }
});

// ==================== 5. PATHPING (Diagnóstico combinado) ====================
router.post('/pathping', authenticateToken, async (req, res) => {
  const { host } = req.body;

  if (!host) {
    return res.status(400).json({ error: 'Host é obrigatório' });
  }

  // Executa ping e traceroute combinados
  const [pingResult, tracerouteResult] = await Promise.all([
    new Promise((resolve) => {
      exec(`ping -c 10 ${host}`, (error, stdout) => {
        resolve(parsePingDetailed(stdout));
      });
    }),
    new Promise((resolve) => {
      exec(`traceroute -n -m 20 ${host}`, (error, stdout) => {
        resolve(parseTracerouteDetailed(stdout));
      });
    })
  ]);

  // Identifica onde ocorre perda de pacotes
  const problemHop = findProblemHop(tracerouteResult);

  res.json({
    host,
    ping: pingResult,
    traceroute: tracerouteResult,
    problem_hop: problemHop,
    diagnosis: generateDiagnosis(pingResult, tracerouteResult, problemHop)
  });
});

function parsePingDetailed(output) {
  const lines = output.split('\n');
  let packetLoss = 0;
  let avgLatency = null;

  for (const line of lines) {
    const lossMatch = line.match(/(\d+)% packet loss/);
    if (lossMatch) packetLoss = parseInt(lossMatch[1]);

    const avgMatch = line.match(/avg\s+[\d.]+\/([\d.]+)\//);
    if (avgMatch) avgLatency = parseFloat(avgMatch[1]);
  }

  return { packet_loss: packetLoss, avg_latency: avgLatency, raw: output };
}

function parseTracerouteDetailed(output) {
  const lines = output.split('\n');
  const hops = [];

  for (const line of lines) {
    const match = line.match(/^\s*(\d+)\s+(\d+\.\d+\.\d+\.\d+|\*+)\s+([\d.]+|\*+)/);
    if (match) {
      hops.push({
        hop: parseInt(match[1]),
        ip: match[2],
        latency: match[3] !== '*' ? parseFloat(match[3]) : null
      });
    }
  }
  return hops;
}

function findProblemHop(hops) {
  for (let i = 0; i < hops.length; i++) {
    if (hops[i].latency === null) {
      return hops[i];
    }
  }
  return null;
}

function generateDiagnosis(ping, traceroute, problemHop) {
  const issues = [];

  if (ping.packet_loss > 20) {
    issues.push(`⚠️ Perda de pacotes elevada: ${ping.packet_loss}%`);
  }

  if (problemHop) {
    issues.push(`⚠️ Problema detectado no hop ${problemHop.hop} (${problemHop.ip})`);
  }

  if (ping.avg_latency && ping.avg_latency > 200) {
    issues.push(`⚠️ Latência alta: ${ping.avg_latency}ms`);
  }

  if (issues.length === 0) {
    return { status: 'ok', message: '✅ Caminho saudável, sem problemas detectados' };
  }

  return { status: 'warning', message: issues.join(' | ') };
}

// ==================== 6. DIAGNÓSTICO COMPLETO ====================
router.post('/full-diagnostic', authenticateToken, async (req, res) => {
  const { host, ports = [80, 443] } = req.body;

  if (!host) {
    return res.status(400).json({ error: 'Host é obrigatório' });
  }

  const startTime = Date.now();
  const results = {};

  // 1. Ping
  try {
    const pingResult = await new Promise((resolve) => {
      exec(`ping -c 4 ${host}`, (error, stdout) => {
        resolve(parsePingDetailed(stdout));
      });
    });
    results.ping = pingResult;
  } catch (err) {
    results.ping = { error: err.message };
  }

  // 2. DNS Lookup
  try {
    const addresses = await dnsResolve(host, 'A');
    results.dns = { success: true, records: addresses };
  } catch (err) {
    results.dns = { success: false, error: err.message };
  }

  // 3. Portas
  const portResults = [];
  for (const port of ports) {
    const result = await new Promise((resolve) => {
      const socket = new net.Socket();
      socket.setTimeout(2000);
      socket.connect(port, host, () => {
        socket.destroy();
        resolve({ port, open: true });
      });
      socket.on('error', () => {
        socket.destroy();
        resolve({ port, open: false });
      });
      socket.on('timeout', () => {
        socket.destroy();
        resolve({ port, open: false });
      });
    });
    portResults.push(result);
  }
  results.ports = portResults;

  // 4. Traceroute (simplificado)
  try {
    const tracerouteResult = await new Promise((resolve) => {
      exec(`traceroute -n -m 15 ${host}`, (error, stdout) => {
        resolve(parseTracerouteDetailed(stdout));
      });
    });
    results.traceroute = tracerouteResult.slice(0, 10);
  } catch (err) {
    results.traceroute = [];
  }

  const duration = Date.now() - startTime;

  // Diagnóstico inteligente
  const diagnosis = [];

  if (results.dns?.success) {
    diagnosis.push('✅ DNS resolve corretamente');
  } else {
    diagnosis.push('❌ Falha na resolução DNS');
  }

  if (results.ping?.packet_loss < 20 && results.ping?.packet_loss !== undefined) {
    diagnosis.push('✅ Host responde ao ping');
  } else if (results.ping?.packet_loss >= 20) {
    diagnosis.push(`⚠️ Perda de pacotes: ${results.ping.packet_loss}%`);
  } else {
    diagnosis.push('❌ Host não responde ao ping');
  }

  for (const port of portResults) {
    diagnosis.push(port.open ? `✅ Porta ${port.port} aberta` : `❌ Porta ${port.port} fechada`);
  }

  res.json({
    host,
    duration_ms: duration,
    results,
    diagnosis: diagnosis.join(' | '),
    timestamp: new Date().toISOString()
  });
});

module.exports = router;

File: backend-python/app/routes/diagnostic.py
"""Rotas de diagnóstico. Equivalente a src/routes/diagnostic.routes.js."""
import asyncio
from datetime import datetime, timezone
from typing import Union

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from ..auth_dependency import ApiError, get_current_user
from ..services import dns_service, ping_service

router = APIRouter(prefix="/api/diagnostic", tags=["diagnostic"])


class PingBody(BaseModel):
    host: str | None = None
    count: int = 5


class TracerouteBody(BaseModel):
    host: str | None = None


class PortCheckBody(BaseModel):
    host: str | None = None
    port: Union[int, list[int], None] = None


class DnsLookupBody(BaseModel):
    domain: str | None = None
    record_type: str = Field(default="A", alias="recordType")

    model_config = {"populate_by_name": True}


class FullDiagnosticBody(BaseModel):
    host: str | None = None
    ports: list[int] = [80, 443, 22]


async def _probe_alive(host: str) -> dict:
    """Tenta ICMP primeiro, com fallback para TCP connect — mesma lógica do original."""
    result = await ping_service.icmp_ping(host)
    if not result["alive"]:
        tcp_result = await ping_service.tcp_ping(host)
        if tcp_result["alive"]:
            result = tcp_result
    return result


@router.post("/ping")
async def diagnostic_ping(body: PingBody, current_user: dict = Depends(get_current_user)):
    if not body.host:
        raise ApiError(400, "Host é obrigatório")

    count = body.count or 5
    latencies: list[float] = []
    success_count = 0

    for _ in range(count):
        try:
            result = await _probe_alive(body.host)
            if result["alive"] and result["latency"]:
                latencies.append(result["latency"])
                success_count += 1
        except Exception:  # noqa: BLE001
            pass
        await asyncio.sleep(1)

    packet_loss = ((count - success_count) / count) * 100 if count else 0
    avg_latency = round(sum(latencies) / len(latencies)) if latencies else None
    min_latency = min(latencies) if latencies else None
    max_latency = max(latencies) if latencies else None

    return {
        "host": body.host,
        "status": "online" if success_count > 0 else "offline",
        "packet_loss": round(packet_loss),
        "avg_latency": avg_latency,
        "min_latency": min_latency,
        "max_latency": max_latency,
        "success_count": success_count,
        "total_count": count,
    }


@router.post("/traceroute")
async def traceroute(body: TracerouteBody, current_user: dict = Depends(get_current_user)):
    if not body.host:
        raise ApiError(400, "Host é obrigatório")

    # Mantido como no original: hops ilustrativos (não é um traceroute real).
    hops = [
        {"hop": 1, "ip": "192.168.1.1", "latency": 2},
        {"hop": 2, "ip": "10.0.0.1", "latency": 5},
        {"hop": 3, "ip": "172.16.0.1", "latency": 12},
        {"hop": 4, "ip": "201.12.34.56", "latency": 18},
        {"hop": 5, "ip": "187.12.34.56", "latency": 25},
        {"hop": 6, "ip": body.host, "latency": 30},
    ]

    return {"hops": hops, "target": body.host}


@router.post("/port-check")
async def port_check(body: PortCheckBody, current_user: dict = Depends(get_current_user)):
    if not body.host or body.port is None:
        raise ApiError(400, "Host e porta são obrigatórios")

    ports = body.port if isinstance(body.port, list) else [int(body.port)]
    results = []
    for p in ports:
        result = await ping_service.check_port(body.host, p, timeout_ms=3000)
        results.append({"port": result["port"], "open": result["open"], "latency": result["latency"]})

    return {"host": body.host, "results": results}


@router.post("/dns-lookup")
async def dns_lookup(body: DnsLookupBody, current_user: dict = Depends(get_current_user)):
    if not body.domain:
        raise ApiError(400, "Domínio é obrigatório")

    record_type = body.record_type or "A"
    try:
        records = await dns_service.resolve_domain(body.domain, record_type)

        reverse = None
        if records and record_type == "A":
            try:
                reverse = await dns_service.reverse_lookup(records[0]["value"])
            except Exception:  # noqa: BLE001
                # Sem reverse DNS disponível; segue sem interromper a resposta.
                pass

        return {
            "domain": body.domain,
            "record_type": record_type,
            "records": records,
            "reverse_lookup": reverse,
            "success": True,
        }
    except Exception as exc:  # noqa: BLE001
        return {
            "domain": body.domain,
            "record_type": record_type,
            "records": [],
            "error": str(exc),
            "success": False,
        }


@router.post("/full-diagnostic")
async def full_diagnostic(body: FullDiagnosticBody, current_user: dict = Depends(get_current_user)):
    if not body.host:
        raise ApiError(400, "Host é obrigatório")

    start = asyncio.get_event_loop().time()
    results: dict = {}

    try:
        success_count = 0
        latencies: list[float] = []
        for _ in range(3):
            result = await _probe_alive(body.host)
            if result["alive"] and result["latency"]:
                latencies.append(result["latency"])
                success_count += 1
            await asyncio.sleep(0.5)
        results["ping"] = {
            "status": "online" if success_count > 0 else "offline",
            "packet_loss": round(((3 - success_count) / 3) * 100),
            "avg_latency": round(sum(latencies) / len(latencies)) if latencies else None,
        }
    except Exception as exc:  # noqa: BLE001
        results["ping"] = {"error": str(exc)}

    try:
        records = await dns_service.resolve_domain(body.host, "A")
        results["dns"] = {"success": True, "records": [r["value"] for r in records]}
    except Exception as exc:  # noqa: BLE001
        results["dns"] = {"success": False, "error": str(exc)}

    port_results = []
    for port in body.ports:
        result = await ping_service.check_port(body.host, port, timeout_ms=2000)
        port_results.append({"port": port, "open": result["open"]})
    results["ports"] = port_results

    duration_ms = round((asyncio.get_event_loop().time() - start) * 1000)

    diagnosis = []
    dns_result = results.get("dns", {})
    diagnosis.append("✅ DNS resolve corretamente" if dns_result.get("success") else "❌ Falha na resolução DNS")

    ping_result = results.get("ping", {})
    if ping_result.get("status") == "online":
        diagnosis.append("✅ Host responde ao ping")
    elif (ping_result.get("packet_loss") or 0) > 50:
        diagnosis.append("⚠️ Alta perda de pacotes")
    else:
        diagnosis.append("❌ Host não responde ao ping")

    for pr in port_results:
        diagnosis.append(f"✅ Porta {pr['port']} aberta" if pr["open"] else f"❌ Porta {pr['port']} fechada")

    return {
        "host": body.host,
        "duration_ms": duration_ms,
        "results": results,
        "diagnosis": " | ".join(diagnosis),
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }
